pkg/kubelet/volume/secret/secret.go

```diff
@@ -0,0 +1,133 @@
+/*
+Copyright 2015 Google Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package secret
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet/volume"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/types"
+	"github.com/golang/glog"
+)
+
+// ProbeVolumePlugin is the entry point for plugin detection in a package.
+func ProbeVolumePlugins() []volume.Plugin {
+	return []volume.Plugin{&secretPlugin{}}
+}
+
+const (
+	secretPluginName = "kubernetes.io/secret"
+)
+
+// secretPlugin implements the VolumePlugin interface.
+type secretPlugin struct {
+	host volume.Host
+}
+
+func (plugin *secretPlugin) Init(host volume.Host) {
+	plugin.host = host
+}
+
+func (plugin *secretPlugin) Name() string {
+	return secretPluginName
+}
+
+func (plugin *secretPlugin) CanSupport(spec *api.Volume) bool {
+	if spec.Source.Secret != nil {
+		return true
+	}
+
+	return false
+}
+
+func (plugin *secretPlugin) NewBuilder(spec *api.Volume, podUID types.UID) (volume.Builder, error) {
+	return plugin.newBuilderInternal(spec, podUID)
+}
+
+func (plugin *secretPlugin) newBuilderInternal(spec *api.Volume, podUID types.UID) (volume.Builder, error) {
+	return &secretVolume{spec.Name, podUID, plugin, &spec.Source.Secret.Target}, nil
+}
+
+func (plugin *secretPlugin) NewCleaner(volName string, podUID types.UID) (volume.Cleaner, error) {
+	return plugin.newCleanerInternal(volName, podUID)
+}
+
+func (plugin *secretPlugin) newCleanerInternal(volName string, podUID types.UID) (volume.Cleaner, error) {
+	return &secretVolume{volName, podUID, plugin, nil}, nil
+}
+
+// secretVolume handles retrieving secrets from the API server
+// and placing them into the volume on the host.
+type secretVolume struct {
+	volName   string
+	podUID    types.UID
+	plugin    *secretPlugin
```

Comment:
Instead of storing a pointer to the plugin in the secretVolume, suggest just storing it in a string in secretVolume and computing it when you initialize secretVolume{} in newCleaner and newBuilder.

This makes it so that secretVolume only has the information it needs. Also, its not immediately obvious that GetPath() would always return the same value, whereas we don't expect the value to change over the lifetime of the Builder/Cleaner objects.